### ISSUE TITLE
Exposing IIIF Endpoint

### ISSUE THREAD

**Issue by rogerhyam** [zenodo/zenodo#1935]
I'm looking at exposing some larger JPGs and possibly other image files stored in Zenodo and to do this I am downloading them, tiling and exposing them as a IIIF endpoint.

In doing this I noticed that Zenodo uses IIIF internally to generate thumbnails and even exposes an Image API endpoint in the links/thumb250 property of the item record. e.g.

https://zenodo.org/api/iiif/v2/f5c4df24-f9c4-4c49-bf8f-54feae0d83dc:3328d0d0-9303-413d-9855-c15d5a26030d:TOLI-22749-EST-01-4-A1-104.jpg/full/250,/0/default.jpg

This might save me a lot of work!

Is there an intention to offer IIIF support for image files stored in Zenodo or would this would be too resource intensive? 

Is the existing IIIF API URL in the record JSON an error? The same size image is available in links/thumbs/250

**Comment by Glignos** [comment]
Currently, Zenodo is indeed exposing an IIIF endpoint.
The links provided to the record under the `thumbs` field correspond to presets for this endpoint, more specifically the width of the thumbnail.
Recently, we also started caching them so this should also speed up the performance.

Using the pure API endpoint is a bit more complicated though since it needs to be contrsucted using information from the metadata of the record and in this case the thumbanail will also not be cached.
Which could result in a heavy load if this endpoint would be 'abused', so if you do use it be cautious since, depending also on the image size, the processing of the picture is a costly operation for us.
I should also mention that the latter is rate limited less generously than the first.

**Comment by jpadfield** [comment]
I have made a proof of concept that pulls in the json details of an uploaded image and then uses the thumb250 value to get all of he required variables for a simple IIIF manifest - https://github.com/jpadfield/iiif-zenodo
I am curious about rate limits and whether directly publishing IIIF manifests would ever be an option for you?

**Comment by jpadfield** [comment]
Just to say the IIIF functionality of Zenodo or the software it runs on InvenioRDM ( https://invenio-software.org/products/rdm/ ) is being improved you can see an initial test example of IIIF previews at: https://tudor-portraits.npg.hasdai.org/records/sr6vd-hwv66

**Comment by abubelinha** [comment]
Thanks @jpadfield 
It looks like those images are stored under a single record id, as in my question. But I think its [api url](https://tudor-portraits.npg.hasdai.org/api/records/sr6vd-hwv66) is not showing IIIF access points like Zenodo does (not even for the first image).

So I understand you just wanted to add some nice examples of the underlying software, but not answering to me (just asking because your comment was made short time after my question).

Unfortunately I am not the kind of user who can run its own repository. My plan is using Zenodo for our images.
But thanks a lot for sharing: that's a very good example of why many images might be shared under a single Zenodo DOI, but still need api-accesible IIIF urls for all of them. This would permit linking to a particular image area of interest (like [this taken from your example](https://tudor-portraits.npg.hasdai.org/api/iiif/v2/sr6vd-hwv66:6768_2008_IRR_mosaic01_adjusted.tif/200,1200,300,800/full/0/default.jpg)).

**Comment by jpadfield** [comment]
Yes, sorry it was not a direct answer to your question, I just wanted to flag that IIIF support was being developed further. Hopefully direct referencing of IIIF content will be easier in the future, but this will probably depend on how each individual repository is setup, I also do hope Zenodo will be able to support this sort of functionality.

**Comment by abubelinha** [comment]
Thanks @jpadfield
As your repository uses the same IIIF software as Zenodo, perhaps you know how persistent are those IIIF urls?
In both of them, I noticed there are some (just guessing) auto-generated ids preceding filenames.

Could they change for example in case of a server restart, or do you think it's safe to link them from external sites?

**Comment by jpadfield** [comment]
I was not directly involved with the setup as it was done by the people developing the software, so I am not sure about that one. It was designed to be persistent, so I would expect the links will persist, if I get any more details I can let you know. However, it was built as a pilot test case, and the software is still being developed, so in this instance things may change. We have been looking to do more work in this area to explore some of these issue a bit more, but we have nothing specific in place yet.

**Comment by rogerhyam** [comment]
@abubelinha I'd be careful of assuming stability of URIs unless advertised as such. Within IIIF I would expect the manifest URI to be stable but none of the others. i.e. the usage pattern would be to store the Manifest URI and get the associated image URIs quite frequently (or at least if they break). We are talking image API URIs here. I presume the only stable URIs in Zenodo are the ones to the metadata record itself (and the DOI wrapper around this URI) and the image ones could change anytime.

I share your ambition to use Zenodo as an image **store** but realistically I don't think they will be open to it being used as an image **server** directly as that commits them unlimited processing as well as storage.

My original vision was to use Zenodo like AWS S3 glacier and have lightweight server somewhere else that pulled files from Zenodo to share through IIIF.

**Comment by rogerhyam** [comment]
IIIF has two main APIs (plus some others). 

The Presentation API provides a manifest that describes a group of canvases that have different things (usually images) painted on them.  Imagine a book with multiple pages with multiple translations and notes attached or a specimen with views from different angles and the parts labelled.

The Image API describes how to get the pixels of any particular image mentioned in a manifest.

The structure is described here

https://iiif.io/get-started/how-iiif-works/

The Manifest API describes the thing we are interested in i.e. the digitise object. So the URI to the manifest is likely to be **more** stable than a URI to an image mentioned in the manifest. The image URI might be generated dynamically just for a particular call to get a manifest (though that is unlikely).

Zenodo uses the image API to deliver thumbnails of files in a repository not the Presentation API as far as I can tell.

You shouldn't assume that any URI will be persistent unless the person who issues it says they intent to try and make it so. Zenodo do not say that for their Image API URIs. 

The web is famously described as "small pieces loosely joined". Therefore all code needs to be written on the basis of "What if this link breaks? What do I fall back to?"

If you are writing code against the Zendo API then you need to only use the calls they document and proceed from there to get the current URI to an image. This is the motivation for my original question. I was asking if they were going to support IIIF (of any form) so that I could use it. I didn't want to hack something based on the fact that I got a hint from the URI structure. The first answer more or less says they can't support it more.

**Comment by abubelinha** [comment]
Thanks for the long explanations. I understood you now.
With all that in consideration, so the assumed answer is "don't use it".  But the image url you linked to more than two years ago is still working.  So the underlying motive for not using it is for not abusing Zenodo servers.

My intention is NOT linking to zenodo image urls (well, I did it above, but only for providing some clickable examples, like you).
I was not meaning to use IIIF server in a direct way, as you suggest (i.e., using it from an IIIF client viewer to browse images). I am more thinking of it as a backup server for images.

So I just want to store those links, but then download, maybe processing (i.e. putting a museum's logo watermark) and -the most important one- **caching** the images using a different service. So I guess this kind of usage would be fair? 
From what you suggest, its better to store just DOIs, and re-check image urls each time you need to download them again to refresh your cache.

My other doubt was about the most proper way of depositing the files at Zenodo: using a shared DOI for many images in a single deposition, or a single deposition and DOI for each image?
I assume the latter would be the preferred choice, so I can use the api for the whole process.

**Comment by rogerhyam** [comment]
We are kind of designing a system here :)

You could use the Zenodo API to make depositions and also retrieve metadata and files. That should be stable but any code would need to be updated when they update the API. The DOI is a fall back. (I don't like DOIs as the give a redundant layer of redirection).

Your strategy for storing images would depend on your needs. There are probably dozens of strategies. You could batch the images into, for example, 100 per zenodo record and link these records into a chain with the most recent record maintaining the most up to date metadata on **all the images in the whole chain**. Each Zenodo record wouldn't change after the metadata was updated to point it to the more recent one in the chain (the files would be immutable). This could grow forever and be persistent.

But I haven't got time to work on this just now. I may return to it later in the year. Good luck.

**Comment by abubelinha** [comment]
Somehow related to this, I would like to have your opinions about which file format would be most appropriate for storage in case I have the choice to convert them before uploading to Zenodo and take profit of IIIF access.
- @rogerhyam said "_I'm looking at exposing some larger JPGs and possibly other image files_"
- @jpadfield show a example with tiffs, but are you using any particular tiff file format? (tiled / pyramidal / compressed)

I have used an [iipimage server](https://github.com/ruven/iipsrv) in the past, and I used to convert my files to a [tiled pyramidal tiff format](https://iipimage.sourceforge.io/documentation/images/).
Now if I plan to use Zenodo, I wonder if that conversion would have any positive effect on speeding up the IIIF access to files (compared to "regular" tiff or jpeg scanned images).  No idea about what Zenodo needs to do internally to provide that IIIF access (perhaps some conversions or calculations which can be harder / slower depending on file formats?)

**Comment by jpadfield** [comment]
The IIIF functionality of InvenioRDM has been developed a lot in the last year - I believe that these developments will slowly be pushed to Zenodo as they update the underlying code. This system will handle the file format for IIIF functionality either Pyramidal Tiff (or Big Tiff) or a version of Jpeg2000 - but that would be their choice based on the server they implement. For you, I would recommend uploading an uncompressed high quality version of the your image designed for long term re-use rather than one tailored for current IIIF servers.
